File: services/identity-service/src/test/java/br/com/deltaglobalbank/identity/features/users/activateUser/ActivateUserTests.java
package br.com.deltaglobalbank.identity.features.users.activateUser;

import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.util.UUID;

import br.com.deltaglobalbank.identity.domain.user.User;
import br.com.deltaglobalbank.identity.domain.user.UserNotFound;
import br.com.deltaglobalbank.identity.domain.user.UserRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class ActivateUserTests {

    private final UserRepository userRepository = mock(UserRepository.class);

    private ActivateUserUseCase useCase;

    @BeforeEach
    void setup() {
        useCase = new ActivateUserUseCase(userRepository);
    }

    @Test
    void activatesUserWhenInSameTenant() {
        UUID userId = UUID.randomUUID();
        UUID tenantId = UUID.randomUUID();
        User user = mock(User.class);
        when(user.getTenantId()).thenReturn(tenantId);
        when(userRepository.findById(userId)).thenReturn(user);
        when(userRepository.save(user)).thenReturn(user);

        useCase.activateUser(userId, tenantId, UUID.randomUUID());

        verify(user, times(1)).activate();
        verify(userRepository, times(1)).save(user);
    }

    @Test
    void throwsWhenUserNotFound() {
        when(userRepository.findById(any())).thenReturn(null);
        assertThrows(UserNotFound.class,
            () -> useCase.activateUser(UUID.randomUUID(), UUID.randomUUID(), UUID.randomUUID()));
    }

    @Test
    void throwsWhenUserBelongsToAnotherTenant() {
        User user = mock(User.class);
        when(user.getTenantId()).thenReturn(UUID.randomUUID());
        when(userRepository.findById(any())).thenReturn(user);
        assertThrows(UserNotFound.class,
            () -> useCase.activateUser(UUID.randomUUID(), UUID.randomUUID(), UUID.randomUUID()));
    }
}

File: services/identity-service/src/test/kotlin/br/com/deltaglobalbank/identity/features/users/activateUser/ActivateUserTests.kt
package br.com.deltaglobalbank.identity.features.users.activateUser

import br.com.deltaglobalbank.identity.domain.user.User
import br.com.deltaglobalbank.identity.domain.user.UserNotFound
import br.com.deltaglobalbank.identity.domain.user.UserRepository
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.mockk
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import io.mockk.verify
import org.springframework.boot.testcontainers.service.connection.ServiceConnection
import org.springframework.context.annotation.Bean
import org.testcontainers.containers.GenericContainer
import org.testcontainers.utility.DockerImageName
import java.util.UUID

@ExtendWith(MockKExtension::class)
class ActivateUserTests {


    @MockK
    lateinit var userRepository: UserRepository

    lateinit var useCase: ActivateUserUseCase

    @BeforeEach
    fun setup() { useCase = ActivateUserUseCase(userRepository) }

    @Bean
    @ServiceConnection(name = "redis")
    fun redisContainer(): GenericContainer<*> =
        GenericContainer(DockerImageName.parse("redis:7-alpine")).withExposedPorts(6379)


    @Test
    fun `activates user when in same tenant`() {
        val userId = UUID.randomUUID()
        val tenantId = UUID.randomUUID()
        val user = mockk<User>(relaxed = true)
        every { user.tenantId } returns tenantId
        every { userRepository.findById(userId) } returns user
        every { userRepository.save(user) } returns user

        useCase.activateUser(userId, tenantId, UUID.randomUUID())

        verify { user.activate() }
        verify { userRepository.save(user) }
    }

    @Test
    fun `throws when user not found`() {
        every { userRepository.findById(any()) } returns null
        assertThrows(UserNotFound::class.java) {
            useCase.activateUser(UUID.randomUUID(), UUID.randomUUID(), UUID.randomUUID())
        }
    }

    @Test
    fun `throws when user belongs to another tenant`() {
        val user = mockk<User>()
        every { user.tenantId } returns UUID.randomUUID()
        every { userRepository.findById(any()) } returns user
        assertThrows(UserNotFound::class.java) {
            useCase.activateUser(UUID.randomUUID(), UUID.randomUUID(), UUID.randomUUID())
        }
    }
}
